File: hellojava/src/main/java/com/fathzer/hellojava/PerftRun.java
package com.fathzer.hellojava;

import com.fathzer.chess.common.PerftResult;

@FunctionalInterface
interface PerftRun {
    record PerfTResultWithDuration(PerftResult result, long duration) {}
    static final com.fathzer.chess.Perft perft = new com.fathzer.chess.Perft();
    static final com.fathzer.chess.optimized.Perft optimizedPerft = new com.fathzer.chess.optimized.Perft();
    PerftResult run(String fen, boolean whitePlaying, int depth);
    private static PerftRun of(boolean optimized) {
        if (optimized) {
            return (fen, whitePlaying, depth) -> optimizedPerft.perft(new com.fathzer.chess.optimized.Board(fen), depth, whitePlaying);
        } else {
            return (fen, whitePlaying, depth) -> perft.perft(new com.fathzer.chess.Board(fen), depth, whitePlaying);
        }
    }
    static PerfTResultWithDuration run(String fen, boolean whitePlaying, int depth, boolean fast) {
        final PerftRun run = of(fast);
        final long start = System.currentTimeMillis();
        final PerftResult result = run.run(fen, whitePlaying, depth);
        return new PerfTResultWithDuration(result, System.currentTimeMillis() - start);
    }
}

File: hellogo/internal/perft/perft_run.go
package perft

import (
	"fmt"
	"time"

	"hellogo/pkg/chess"
	"hellogo/pkg/chess/optimized"
	chesscommon "hellogo/pkg/chess/common"
)

// ResultWithDuration holds the result of a Perft test along with its execution duration
type ResultWithDuration struct {
	Result   *chesscommon.PerftResult
	Duration time.Duration
}

// Runner defines the interface for running Perft tests
type Runner interface {
	run(fen string, whitePlaying bool, depth int) *chesscommon.PerftResult
}

type standardPerft struct{}
type optimizedPerft struct{}

var (
	standardPerftInstance = &standardPerft{}
	optimizedPerftInstance = &optimizedPerft{}
)

// Run executes a Perft test with the given parameters and returns the result with timing
func Run(fen string, whitePlaying bool, depth int, useOptimized bool) (ResultWithDuration, error) {
	var runner Runner
	if useOptimized {
		runner = optimizedPerftInstance
	} else {
		runner = standardPerftInstance
	}

	start := time.Now()
	result := runner.run(fen, whitePlaying, depth)
	elapsed := time.Since(start)

	return ResultWithDuration{
		Result:   result,
		Duration: elapsed,
	}, nil
}

func (p *standardPerft) run(fen string, whitePlaying bool, depth int) *chesscommon.PerftResult {
	board, err := chess.NewBoard(fen)
	if err != nil {
		panic(fmt.Sprintf("invalid FEN: %v", err))
	}
	perft := chess.NewPerft()
	// Using NonBulk as the default type to match Java implementation
	result, _ := perft.PerftWithType(board, depth, chesscommon.NonBulk, whitePlaying)
	return result
}

func (p *optimizedPerft) run(fen string, whitePlaying bool, depth int) *chesscommon.PerftResult {
	board := optimized.NewBoard(fen)
	perft := optimized.NewPerft()
	// Using NonBulk as the default type to match Java implementation
	return perft.PerftWithType(board, depth, chesscommon.NonBulk, whitePlaying)
}
